spikeinterface/sortingcomponents/peak_pipeline.py

```diff
@@ -11,88 +11,136 @@
-            self.contact_locations = recording.get_channel_locations()
-            self.channel_distance = get_channel_distances(recording)
-            self.neighbours_mask = self.channel_distance < local_radius_um
+    This is a generic object that will make some computation on peak given a buffer
+    of traces.
+    Typically used for extrating features (amplitudes, localization, ...)
+    
+    A Node can optionally connect to other nodes with the parents and receive inputs from others.
+    """
+    def __init__(self, recording, name, have_global_output, parents=None):
+        self.recording = recording
+        self.name = name
+        self.have_global_output = have_global_output
+        if isinstance(parents, str):
+            # only one parents is allowed
+            parents = [parents]
+        self.parents = parents
+        
+        self._kwargs = dict(
+            name=name,
+            have_global_output=have_global_output,
+        )
+        if parents is not None:
+            self._kwargs['parents'] = parents
+        
+        self.other_nodes = None
 
     @classmethod
     def from_dict(cls, recording, kwargs):
         return cls(recording, **kwargs)
 
     def to_dict(self):
         return self._kwargs
-
+        
+    def give_other_nodes(self, other_nodes):
+        # other 
+        self.other_nodes = other_nodes
+    
+    def post_check(self):
+        # can optionaly be overwritten
+        # this can trigger a check for compatibility with other nodes (typically parents)
+        pass
+    
     def get_trace_margin(self):
         # can optionaly be overwritten
-        if self.need_waveforms:
-            return max(self.nbefore, self.nafter)
-        elif self.nbefore is not None:
-            return max(self.nbefore, self.nafter)
-        else:
-            return 0
+        return 0
 
     def get_dtype(self):
         raise NotImplementedError
 
-    def compute_buffer(self, traces, peaks, waveforms=None):
-        raise NotImplementedError
 
+class ExtractDenseWaveforms(PipelineNode):
+    def __init__(self, recording, name='extract_dense_waveforms', have_global_output=False,
+                         ms_before=None, ms_after=None):
+        PipelineNode.__init__(self, recording, name=name, have_global_output=have_global_output)
+
+        self.nbefore = int(ms_before * recording.get_sampling_frequency() / 1000.)
+        self.nafter = int(ms_after * recording.get_sampling_frequency() / 1000.)
+        
+        self._kwargs['ms_before'] = float(ms_before)
+        self._kwargs['ms_after'] = float(ms_after)
+
+    def get_trace_margin(self):
+        return max(self.nbefore, self.nafter)
+    
+    def compute(self, traces, peaks):
+        waveforms = traces[peaks['sample_ind'][:, None] + np.arange(-self.nbefore, self.nafter)]
+        return waveforms
 
-def run_peak_pipeline(recording, peaks, steps, job_kwargs, job_name='peak_pipeline', squeeze_output=True):
+
+def check_graph(nodes):
```

Comment:
I think this method could be separated in two:

`assert_graph_structure` which can be run in the general init and `propagate_node_instances` (or something like that) that is used to propagate the instances in the `worker_ctx` intializer.

This is probably slightly more efficient and more readable as well (the function right now does not have a specific enough name and is performing two roles).

Reply:
OK thank you. This make sens. I will do that.

Reply:
done